File: Solution/322.零钱兑换/322.零钱兑换.cpp
/*
 * @Descroption: LeetCode 322. 零钱兑换
 * @Author: EmoryHuang
 * @Date: 2021-05-20 16:20:51
 * @解题思路:
 * 动态规划问题，背包问题
 * dp[i] 表示组成金额为 i 的最少硬币个数
 * 状态转移方程：dp[i] = min(dp[i], dp[i - coin] + 1)
 * 由于需要金额恰好为 amount，因此初始化时将dp[0]=0外，其他均初始化为INT_MAX
 */

class Solution {
   public:
    int coinChange(vector<int>& coins, int amount) {
        int n = coins.size();
        vector<int> dp(amount + 1, 0x3f3f3f3f);  // 初始化为INT_MAX
        dp[0] = 0;
        for (int i = 1; i <= amount; i++) {  // 遍历金额
            for (int j = 0; j < n; j++) {
                if (coins[j] <= i) 
                    dp[i] = min(dp[i], dp[i - coins[j]] + 1);
            }
        }
        return dp[amount] != 0x3f3f3f3f ? dp[amount] : -1;
    }
};

File: Solution/322.零钱兑换/322.零钱兑换.java
/*
 * @Description: LeetCode 322. 零钱兑换
 * @Author: EmoryHuang
 * @Date: 2023-08-16 10:26:58
 * @解题思路:
 * 动态规划，背包问题
 * dp[i] 表示组成和为 i 的零钱的最少数量
 * dp[i] = min(dp[i], dp[i - coins[j]] + 1)
 */

import java.util.*;

class Solution {
    public int coinChange(int[] coins, int amount) {
        int[] dp = new int[amount + 1];
        Arrays.fill(dp, 0x3f3f3f3f);
        dp[0] = 0;
        for (int i = 0; i < coins.length; i++) {
            for (int j = coins[i]; j <= amount; j++) {
                dp[j] = Math.min(dp[j], dp[j - coins[i]] + 1);
            }
        }
        return dp[amount] == 0x3f3f3f3f ? -1 : dp[amount];
    }
}